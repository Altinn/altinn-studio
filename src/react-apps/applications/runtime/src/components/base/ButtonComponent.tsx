import * as React from 'react';
import { connect } from 'react-redux';
import { makeGetValidationsSelector } from '../../selectors/getValidations';
import { IValidations } from '../../types/global';
import { canFormBeSaved, getErrorCount } from '../../utils/validation';
import FormDataActions from './../../features/form/data/actions/index';
import { IAltinnWindow, IRuntimeState } from './../../types';
export interface IButtonProvidedProps {
  id: string;
  text: string;
  disabled: boolean;
  handleDataChange: (value: any) => void;
  unsavedChanges: boolean;
  formDataCount: number;
}

export interface IButtonProps extends IButtonProvidedProps {
  validations: IValidations;
}

export interface IButtonState { }

export class ButtonComponentClass extends React.Component<IButtonProps, IButtonState> {
  public renderSubmitButton = () => {
    const disabled = (getErrorCount(this.props.validations) > 0) || this.props.unsavedChanges;
    return (
      <button
        type='submit'
        className={disabled ? 'a-btn a-btn-success disabled' : 'a-btn a-btn-success'}
        onClick={this.submitForm}
        disabled={disabled}
        id={this.props.id}
        style={{ marginBottom: '0' }}
      >
        {this.props.text}
      </button>
    );
  }

  // TODO: Remove saveButton and functions (and sagas) when we have implemented automatic save.
  public renderSaveButton = () => {
    const disabled = !this.props.unsavedChanges || !canFormBeSaved(this.props.validations);
    return (
      <button
        type='submit'
        className={disabled ?
          'a-btn a-btn-success disabled' : 'a-btn a-btn-success'}
        onClick={this.saveFormData}
        disabled={disabled}
        id='saveBtn'
        style={{ marginBottom: '0' }}
      >
        Lagre
      </button>
    );
  }

  public saveFormData = () => {
    const { org, app, instanceId } = window as Window as IAltinnWindow;
    FormDataActions.submitFormData(
      `${window.location.origin}/${org}/${app}/api/${instanceId}`,
    );
  }

  public submitForm = () => {
    const {org, app, instanceId } = window as Window as IAltinnWindow;
    FormDataActions.submitFormData(
      `${window.location.origin}/${org}/${app}/api/${instanceId}`,
      'Complete',
    );
  }

  public render() {
    return (
      <div className='row'>
        <div className='a-btn-group' style={{ marginTop: '3.6rem', marginBottom: '0' }}>
          {this.renderSaveButton()}
          {this.renderSubmitButton()}
        </div>
      </div>
    );
  }
}

const makeMapStateToProps = () => {
  const GetValidations = makeGetValidationsSelector();
  const mapStateToProps = (state: IRuntimeState, props: IButtonProvidedProps): IButtonProps => {
    return {
      validations: GetValidations(state),
      ...props,
    };
  };
  return mapStateToProps;
};

export const ButtonComponent = connect(makeMapStateToProps)(ButtonComponentClass);
