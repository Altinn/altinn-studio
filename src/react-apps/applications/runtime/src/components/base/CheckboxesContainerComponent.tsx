import { FormControlLabel, FormGroup } from '@material-ui/core';
import Checkbox, { CheckboxProps } from '@material-ui/core/Checkbox';
import FormControl from '@material-ui/core/FormControl';
import { makeStyles } from '@material-ui/core/styles';
import classNames = require('classnames');
import * as React from 'react';
import altinnTheme from '../../../../shared/src/theme/altinnAppTheme';
import { renderValidationMessagesForComponent } from '../../utils/render';

export interface ICheckboxContainerProps {
  id: string;
  formData: any;
  handleDataChange: (value: any) => void;
  handleFocusUpdate: (value: any) => void;
  isValid: boolean;
  validationMessages: any;
  options: any[];
  preselectedOptionIndex: number;
  readOnly: boolean;
  shouldFocus: boolean;
}

const useStyles = makeStyles({
  root: {
    '&:hover': {
      backgroundColor: 'transparent',
    },
  },
  icon: {
    'border': '2px solid #1EAEF7',
    'width': 24,
    'height': 24,
    'backgroundColor': '#ffffff',
    '$root.Mui-focusVisible &': {
      outline: '2px solid #ff0000',
      outlineOffset: 0,
      outlineColor: altinnTheme.altinnPalette.primary.blueDark,
    },
    'input:hover ~ &': {
      borderColor: altinnTheme.altinnPalette.primary.blueDark,
    },
    'input:disabled ~ &': {
      boxShadow: 'none',
      background: 'rgba(206,217,224,.5)',
    },
  },
  checkedIcon: {
    'backgroundColor': '#ffffff',
    '&:before': {
      display: 'block',
      width: 20,
      height: 20,
      // tslint:disable-next-line: max-line-length
      backgroundImage: "url(\"data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 8 8'%3E%3Cpath fill='%23000000' d='M6.564.75l-3.59 3.612-1.538-1.55L0 4.26 2.974 7.25 8 2.193z'/%3E%3C/svg%3E\")",
      content: '""',
    },
    'input:hover ~ &': {
      borderColor: altinnTheme.altinnPalette.primary.blueDark,
    },
  },
});

export const CheckboxContainerComponent = (props: ICheckboxContainerProps) => {
  const classes = useStyles(props);

  const [selected, setSelected] = React.useState([]);
  const checkBoxesIsRow: boolean = (props.options.length <= 2);

  React.useEffect(() => {
    returnState();
  }, [props]);

  const returnState = () => {
    if (
      !props.formData &&
      props.preselectedOptionIndex &&
      props.options &&
      props.preselectedOptionIndex < props.options.length
    ) {
      const preSelected: string[] = [];
      preSelected[props.preselectedOptionIndex] = props.options[props.preselectedOptionIndex].value;
      setSelected(preSelected);
    } else {
      setSelected(props.formData ? props.formData.split(',') : []);
    }
  };

  const onDataChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newSelected = selected;

    if (newSelected[event.target.value] === event.target.name) {
      newSelected[event.target.value] = '';
    } else {
      newSelected[event.target.value] = event.target.name;
    }

    setSelected(newSelected);
    props.handleDataChange(selectedHasValues(newSelected) ? newSelected.join() : '');
  };

  const selectedHasValues = (select: string[]): boolean => {
    return select.some((element) => element !== '');
  };

  const isOptionSelected = (option: string) => {
    return selected.indexOf(option) > -1;
  };

  const StyledCheckbox = (checkboxProps: CheckboxProps) => {
    return (
      <Checkbox
        className={classes.root}
        disableRipple={true}
        color='default'
        checkedIcon={<span className={classNames(classes.icon, classes.checkedIcon)} />}
        icon={<span className={classes.icon} />}
        inputProps={{ 'aria-label': 'decorative checkbox' }}
        {...checkboxProps}
      />
    );
  };

  return(
    <FormControl>
      <FormGroup row={checkBoxesIsRow}>
        {props.options.map((option, index) => (
          <React.Fragment key={index}>
            <FormControlLabel
              key={index}
              control={(
                <StyledCheckbox
                  checked={isOptionSelected(option.value)}
                  onChange={onDataChanged}
                  value={index}
                  name={option.value}
                />
              )}
              label={option.label}
            />
            { props.validationMessages &&
              this.isOptionSelected(option.value) &&
              renderValidationMessagesForComponent(props.validationMessages.simpleBinding, props.id) }
          </React.Fragment>
        ))}
      </FormGroup>
    </FormControl>
  );

};
