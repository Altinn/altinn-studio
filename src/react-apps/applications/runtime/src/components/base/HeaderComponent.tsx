import * as React from 'react';

export interface IHeaderProps {
  id: string;
  text: string;
  size?: string;
}

export interface IHeaderState { }

export class HeaderComponent extends React.Component<IHeaderProps, IHeaderState> {

  public h2style = {
    marginTop: '4.8rem',
    marginBottom: '0',
  };

  public h3style = {
    marginTop: '4.8rem',
    marginBottom: '0',
  };

  public h4style = {
    marginTop: '4.8rem',
    marginBottom: '0',
  };

  public renderHeader(): JSX.Element {
    switch (this.props.size) {
      case ('S'): {
        return <h4 id={this.props.id} style={this.h4style}>{this.props.text}</h4>;
      }

      case ('M'): {
        return <h3 id={this.props.id} style={this.h3style}>{this.props.text}</h3>;
      }

      case ('L'): {
        return <h2 id={this.props.id} style={this.h2style}>{this.props.text}</h2>;
      }

      default: {
        return <h4 id={this.props.id}>{this.props.text}</h4>;
      }
    }
  }

  public render() {
    return (
      this.renderHeader()
    );
  }
}
