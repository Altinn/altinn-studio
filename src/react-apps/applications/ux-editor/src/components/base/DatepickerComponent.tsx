import * as React from 'react';
import '../../styles/InputComponent.css';

export interface IDatepickerProps {
  id: string;
  component: IFormComponent;
  formData: any;
  handleDataChange: (value: any) => void;
  isValid?: boolean;
}

export interface IDatePickerState {
  value: string;
}

export class DatepickerComponent
  extends React.Component<IDatepickerProps, IDatePickerState> {

  constructor(_props: IDatepickerProps, _state: IDatePickerState) {
    super(_props, _state);
    this.state = {
      value: _props.formData ? _props.formData : '',
    };
  }

  public onDataChanged = (e: any) => {
    this.setState({
      value: e.target.value,
    });
  }

  public onDataChangeSubmit = () => {
    this.props.handleDataChange(this.state.value);
  }

  public render() {
    return (
      <input
        id={this.props.id}
        type={this.props.component.type}
        onBlur={this.onDataChangeSubmit}
        onChange={this.onDataChanged}
        disabled={this.props.component.disabled}
        required={this.props.component.required}
        className={this.props.isValid ? 'form-control' : 'form-control validation-error'}
        value={this.state.value}
      />
    );
  }
}
