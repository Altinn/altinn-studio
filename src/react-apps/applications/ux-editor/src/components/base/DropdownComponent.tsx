import * as React from 'react';

export interface IDropdownProps {
  id: string;
  component: IFormComponent;
  formData: any;
  handleDataChange: (value: any) => void;
  getTextResource: (resourceKey: string) => string;
  isValid?: boolean;
  designMode: boolean;
}

export interface IDropdownState {
  title: string;
  options: IOptions[];
  name: string;
}

export class DropdownComponent
  extends React.Component<IDropdownProps, IDropdownState> {

  public onDataChanged = (e: any) => {
    this.props.handleDataChange(e.target.value);
  }

  public render() {
    const { options } = this.props.component;
    return (
      <select
        id={this.props.id}
        value={this.props.formData}
        className={this.props.isValid ? 'custom-select a-custom-select'
        : 'custom-select a-custom-select validation-error'}
        onChange={this.onDataChanged}
      >
        {options.map((option, index) => (
          <option key={index} value={option.value}>
            {this.props.designMode ? option.label : this.props.getTextResource(option.label)}
          </option>
        ))}
      </select>
    );
  }
}
