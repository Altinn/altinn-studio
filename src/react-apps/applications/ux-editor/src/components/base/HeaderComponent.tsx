import * as React from 'react';

export interface IHeaderProps {
  component: IFormComponent;
  text: string;
  size?: string;
}

export interface IHeaderState { }

export class HeaderComponent extends React.Component<IHeaderProps, IHeaderState> {

  public renderHeader(): JSX.Element {
    switch (this.props.size) {
      case ('S'): {
        return <h4 className='a-sectionSubTitle' id={this.props.component.id}>{this.props.text}</h4>;
      }

      case ('M'): {
        return <h3 className='a-sectionSubTitle' id={this.props.component.id}>{this.props.text}</h3>;
      }

      default: {
        return <h2 className='a-sectionTitle' id={this.props.component.id}>{this.props.text}</h2>;
      }
    }
  }

  public render() {
    return (
      this.renderHeader()
    );
  }
}
