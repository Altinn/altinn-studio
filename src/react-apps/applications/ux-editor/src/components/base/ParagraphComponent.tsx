import * as React from 'react';

export interface IParagraphProps {
  component: IFormComponent;
  text: string;
}

export interface IParagraphState { }

export class ParagraphComponent extends React.Component<IParagraphProps, IParagraphState> {

  public render() {
    return (
      <span id={this.props.component.id}>{this.props.text}</span>
    );
  }
}
