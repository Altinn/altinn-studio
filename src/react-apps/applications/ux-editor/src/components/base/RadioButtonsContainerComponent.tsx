import * as React from 'react';

export interface IRadioButtonsContainerProps {
  id: string;
  component: IFormComponent;
  formData: any;
  handleDataChange: (value: any) => void;
  getTextResource: (resourceKey: string) => string;
  isValid?: boolean;
  designMode: boolean;
}

export interface IRadioButtonsContainerState {
  selected: any;
}

export class RadioButtonContainerComponent
  extends React.Component<
  IRadioButtonsContainerProps,
  IRadioButtonsContainerState
  > {
  constructor(props: IRadioButtonsContainerProps, state: IRadioButtonsContainerState) {
    super(props, state);
    this.state = {
      selected: '',
    };
  }

  public onDataChanged = (selectedValue: any) => {
    this.setState({
      selected: selectedValue,
    });
    this.props.handleDataChange(selectedValue);
  }

  public render() {
    const { options } = this.props.component;
    return (
      <div className={this.props.isValid ? 'form-group' : 'form-group validation-error'} id={this.props.id}>
        {options.map((option, index) => (
          <div
            className='custom-control custom-radio pl-0 a-custom-radio custom-control-stacked'
            key={index}
            onClick={this.onDataChanged.bind(this, option.value)}
          >
            <input
              type='radio'
              name={'radio-' + this.props.id}
              className='custom-control-input'
              checked={this.state.selected === option.value}
            />
            <label className='custom-control-label pl-3 a-radioButtons-title'>
              {this.props.designMode ? option.label : this.props.getTextResource(option.label)}
            </label>
          </div>
        ))}
      </div>
    );
  }
}
