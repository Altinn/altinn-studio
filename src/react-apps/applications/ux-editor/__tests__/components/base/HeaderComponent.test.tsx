import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { mount } from 'enzyme';

import { HeaderComponent } from '../../../src/components/base/HeaderComponent';

describe('>>> components/base/HeaderComponent.tsx --- Snapshot', () => {
  let mockComponent: any;
  let mockText: string;
  let mockSize: string;

  beforeEach(() => {
    mockComponent = {
      id: 'test-id',
      title: 'test-headercomponent',
      component: 'Header',
    };
    mockText = 'test';
    mockSize = 'S';
  });

  it('+++ should match snapshot', () => {
    const rendered = renderer.create(
      <HeaderComponent
        component={mockComponent}
        text={mockText}
        size={mockSize}
      />,
    );
    expect(rendered).toMatchSnapshot();
  });

  it('+++ should render <h2> if size is not \'S\' or \'M\'', () => {
    const wrapper = mount(
      <HeaderComponent
        component={mockComponent}
        text={mockText}
      />,
    );
    expect(wrapper.contains(<h2 className='a-sectionTitle' id={'test-id'}>test</h2>)).toBe(true);
  });

  it('+++ should render <h3> if size is \'M\'', () => {
    const wrapper = mount(
      <HeaderComponent
        component={mockComponent}
        text={mockText}
        size={'M'}
      />,
    );
    expect(wrapper.contains(<h3 className='a-sectionSubTitle' id={'test-id'}>test</h3>)).toBe(true);
  });

  it('+++ should render <h3> if size is \'S\'', () => {
    const wrapper = mount(
      <HeaderComponent
        component={mockComponent}
        text={mockText}
        size={'S'}
      />,
    );
    expect(wrapper.contains(<h4 className='a-sectionSubTitle' id={'test-id'}>test</h4>)).toBe(true);
  });
});
