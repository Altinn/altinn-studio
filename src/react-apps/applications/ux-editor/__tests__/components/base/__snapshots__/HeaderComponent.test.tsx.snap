// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> components/base/HeaderComponent.tsx --- Snapshot +++ should match snapshot 1`] = `
<h4
  className="a-sectionSubTitle"
  id="test-id"
>
  test
</h4>
`;
